File: EmpolyeeImportance.py
//BFS Approcah
//Time Complexity: O(n)
//Space Complexity:O(n)
// Did the code run successfully on leet code: Yes

"""
# Definition for Employee.
class Employee(object):
    def __init__(self, id, importance, subordinates):
    	#################
        :type id: int
        :type importance: int
        :type subordinates: List[int]
        #################
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""

class Solution(object):
    def getImportance(self, employees, id):
        """
        :type employees: List[Employee]
        :type id: int
        :rtype: int
        """
        res = 0
        hashmap = {}
        for e in employees:
            # print(e.id)
            hashmap[e.id] = e
        q = deque([id])
        # print(hashmap)
        # print(q)
        
        # print(len(q))
        while(len(q)):
                 
                currid = q.popleft()
                print(currid)
                node = hashmap[currid]
                res = res+node.importance
                print(res)
                for i in range(len(node.subordinates)):
                    q.append(node.subordinates[i])
        return res           
                

File: empolyeeImportance.java
// Time Complexity : o(V + E)
// Space Complexity : o(V + E)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
// Leetcode : 690
//Approach : BFS



/*
// Definition for Employee.
class Employee {
    public int id;
    public int importance;
    public List<Integer> subordinates;
};
*/


class Solution {
    public int getImportance(List<Employee> employees, int id) {
        //base
        if(employees == null || employees.size() == 0) return 0;
    
        HashMap<Integer, Employee> employee = new HashMap<>();
        for(Employee e: employees ){
            employee.put(e.id ,e);
        }
        
        Queue<Integer> q = new LinkedList<>();
        q.add(id);
        int result = 0;
        while(!q.isEmpty()){
             int eid = q.poll();
             Employee e = employee.get(eid);
             result += e.importance;
             for(int sid : e.subordinates){
                 q.add(sid);
             }    
        }
    return result;
        
    }
}


// Time Complexity : o(V + E)
// Space Complexity : o(V + E)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
// Leetcode : 994
//Approach : DFS

class Solution {
    int result;
    HashMap<Integer, Employee> employee = new HashMap<>();
    public int getImportance(List<Employee> employees, int id) {
        //base
        if(employees == null || employees.size() == 0) return 0;     
        for(Employee e: employees ){
            employee.put(e.id ,e);
        }
        dfs(id);
  return result;        
    }
    
            private void dfs(int id){
            
            //base
            //No base required
            //logic
            Employee e = employee.get(id);
            result+= e.importance;
            for(int sid: e.subordinates){
                dfs(sid);
            }   
        }
}